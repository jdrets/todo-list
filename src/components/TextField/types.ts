export interface TextFieldTypes {
  placeholder?: string
  onChange: () => void
  label?: string
  value?: string | null
  fullWidth?: boolean
  name?: string
  required?: boolean
  multiline?: boolean
  type?: string
}
