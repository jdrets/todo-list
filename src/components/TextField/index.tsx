import React, { FunctionComponent } from 'react'

import {
  Wrapper,
  Input,
  Label,
  RequiredLabel
} from './styles'
import { TextFieldTypes } from './types'

const TextField: FunctionComponent<TextFieldTypes> = ({
  placeholder,
  onChange,
  label,
  value,
  fullWidth,
  name,
  required,
  multiline
}) => (
  <Wrapper>
    <Label>
      {label} {required && <RequiredLabel>*</RequiredLabel>}
    </Label>
    <Input
      as={multiline ? 'textarea' : 'input'}
      value={value}
      placeholder={placeholder}
      onChange={onChange}
      fullWidth={fullWidth}
      multiline={multiline}
      name={name}
      required
    />
  </Wrapper>
)

export default TextField
