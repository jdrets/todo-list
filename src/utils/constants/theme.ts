const palette = {
  purpleDark: '#4B00BC',
  purple: '#6600FF',
  purpleMedium: '#8533FF',
  purpleLight: '#E3D0FF',
  green: '#59BC6F',
  greenLight: '#E6F5E2',
  red: '#EB5757',
  redLight: '#FCE4E4',
  yellow: '#DD9900',
  yellowLight: '#FEF1D8',
  white: '#FFFFFF',
  black: '#000000',
  gray: '#898989',
  grayLight: '#E0E0E0'
}

const theme = {
  colors: {
    primaryColor: palette.purple,
    primaryColorMedium: palette.purpleMedium,
    primaryColorLight: palette.purpleLight,
    primaryColorDark: palette.purpleDark,
    ...palette
  },
  layout: {
    breakpoints: {
      medium: '768px'
    },
    mobile: {
      width: '100%',
      padding: '0 24px'
    },
    desktop: {
      width: '1050px'
    }
  }
}

export default theme
