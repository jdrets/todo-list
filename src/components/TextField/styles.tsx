import styled from 'styled-components'

import theme from '../../utils/constants/theme'
import { TextFieldTypes } from './types'

export const Wrapper = styled.div`
  display: block;
`

export const Input = styled.input<TextFieldTypes>`
  background-color: white;
  border-radius: 4px;
  border: 1px solid ${theme.colors.grayLight};
  height: ${props => props.multiline && '150px'};
  outline: none;
  padding: 12px;
  resize: none;
  transition: 0.3s;
  width: ${props => props.fullWidth ? '100%' : 'auto'};
  
  &:hover {
    border-color: ${theme.colors.primaryColorLight};
  }
  
  &:focus {
    border-color: ${theme.colors.primaryColor};
  }
  
  &::-webkit-calendar-picker-indicator {
    opacity: 0.2;
    padding: 0;
  }
`

export const Label = styled.div`
  font-size: 14px;
  margin-bottom: 4px;
  font-weight: 300;
`

export const RequiredLabel = styled.span`
  color: ${theme.colors.red};
`
