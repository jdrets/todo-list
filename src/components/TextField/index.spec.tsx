import React from 'react'
import { ThemeProvider } from 'styled-components'

import { render, fireEvent } from '@testing-library/react'

import TextField from '.'
import theme from '../../utils/constants/theme'

const setup = (mockData = {}) => {
  return render(
    <ThemeProvider theme={theme}>
      <TextField
        placeholder="dummy placeholder"
        onChange={jest.fn()}
        label="dummy label"
        value="dummy value"
        fullWidth
        name="dummy name"
        required
        multiline
        {...mockData}
      />
    </ThemeProvider>
  )
}

describe('<TextField />', () => {
  test('should match with snapshot', () => {
    const { container } = setup()

    expect(container).toMatchSnapshot()
  })

  test('component tag should be a input when multiline is null', () => {
    setup({
      multiline: null
    })

    const input = document.getElementsByTagName('input')[0]
    const textarea = document.getElementsByTagName('textarea')[0]

    expect(input).toBeInTheDocument()
    expect(textarea).toBeFalsy()
  })

  test('should call onChange function when input is change', () => {
    const onChangeMock = jest.fn()
    const { getByText } = setup({
      onChange: onChangeMock
    })
    const input = getByText('dummy value')

    fireEvent.change(input, {
      target: {
        value: 'new value'
      }
    })

    expect(onChangeMock).toHaveBeenCalled()
  })

  test('should render required label when required prop is true', () => {
    const { getByText } = setup({
      required: true
    })
    const requiredLabel = getByText('*')

    expect(requiredLabel).toBeInTheDocument()
  })

  test('should not render required labelasd when required prop is null', () => {
    const { queryByText } = setup({
      required: false
    })

    expect(queryByText('*')).toBeNull()
  })
})
