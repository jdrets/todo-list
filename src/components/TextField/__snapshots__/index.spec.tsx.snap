// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should match with snapshot 1`] = `
<div>
  <div
    class="sc-bdnylx lmgoAG"
  >
    <div
      class="sc-dlnjPT iSoqbz"
    >
      dummy label
       
      <span
        class="sc-hKFyIo blOphI"
      >
        *
      </span>
    </div>
    <textarea
      class="sc-gtssRu YlOWv"
      name="dummy name"
      placeholder="dummy placeholder"
      required=""
    >
      dummy value
    </textarea>
  </div>
</div>
`;
